sdk/compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Upgrade.hs

```diff
@@ -8,147 +8,274 @@ module DA.Daml.LF.TypeChecker.Upgrade (
+    when shouldTypecheck $ do
+      case mbUpgradedPkg of
+        Nothing -> pure ()
+        Just (upgradedPkg, upgradingDeps) -> do
+            deps <- checkUpgradeDependenciesM deps upgradingDeps
+            checkUpgradeBoth Nothing pkg (upgradedPkg, deps)
+      checkUpgradeSingle Nothing pkg
+
+checkUpgradeBoth :: Maybe Context -> LF.Package -> ((LF.PackageId, LF.Package), UpgradingDeps) -> TcMF (Version, UpgradeInfo) ()
+checkUpgradeBoth mbContext pkg ((upgradedPkgId, upgradedPkg), upgradingDeps) =
+  let presentWorld = initWorldSelf [] pkg
+      pastWorld = initWorldSelf [] upgradedPkg
+      upgradingWorld = Upgrading { _past = pastWorld, _present = presentWorld }
+      withMbContext :: TcUpgradeM () -> TcUpgradeM ()
+      withMbContext =
+        case mbContext of
+          Nothing -> id
+          Just context -> withContextF present' context
+  in
+  withReaderT (\(version, upgradeInfo) -> UpgradingEnv (mkGamma version upgradeInfo <$> upgradingWorld) upgradingDeps) $
+    withMbContext $
+      checkUpgradeM (UpgradedPackageId upgradedPkgId) (Upgrading upgradedPkg pkg)
+
+checkUpgradeSingle :: Maybe Context -> LF.Package -> TcMF (Version, UpgradeInfo) ()
+checkUpgradeSingle mbContext pkg =
+  let presentWorld = initWorldSelf [] pkg
+      withMbContext :: TcM () -> TcM ()
+      withMbContext = maybe id withContext mbContext
+  in
+  withReaderT (\(version, upgradeInfo) -> mkGamma version upgradeInfo presentWorld) $
+    withMbContext $ do
+      checkNewInterfacesAreUnused pkg
+      checkNewInterfacesHaveNoTemplates pkg
 
 checkModule
   :: LF.World -> LF.Module
-  -> Version -> UpgradeInfo
-  -> Maybe (LF.PackageId, LF.Package)
+  -> [LF.DalfPackage] -> Version -> UpgradeInfo
+  -> Maybe ((LF.PackageId, LF.Package), [(LF.PackageId, LF.Package)])
   -> [Diagnostic]
-checkModule world0 module_ =
-    let world = extendWorldSelf module_ world0
-        checkBoth upgradedPkgId upgradedPkg =
-          case NM.lookup (NM.name module_) (LF.packageModules upgradedPkg) of
-            Nothing -> pure ()
-            Just pastModule ->
-              let upgradingModule = Upgrading { _past = pastModule, _present = module_ }
-              in
-              checkModuleM upgradedPkgId upgradingModule
-        checkSingle = do
-          checkNewInterfacesAreUnused module_
-          checkNewInterfacesHaveNoTemplates module_
-    in
-    checkBothAndSingle world checkBoth checkSingle
+checkModule world0 module_ deps version upgradeInfo mbUpgradedPkg =
+  extractDiagnostics version upgradeInfo $
+    when (shouldTypecheck version upgradeInfo) $ do
+      let world = extendWorldSelf module_ world0
+      case mbUpgradedPkg of
+        Nothing -> pure ()
+        Just ((upgradedPkgIdRaw, upgradedPkg), upgradingDeps) -> do
+            let upgradedPkgId = UpgradedPackageId upgradedPkgIdRaw
+            deps <- checkUpgradeDependenciesM deps upgradingDeps -- TODO: Check if this causes quadratic blowup of dep checks
```

Comment:
I think it might, but also I think you want to topo sort all packages based on the "depends on" relationship at toplevel and treat them in that order. (See comment below about comparing type constructors.)

Reply:
As discussed, the map from package name to package IDs already checked for upgrades could be a shake rule to prevent recomputation and avoid the O(n*m) complexity.

Reply:
I've fixed upgradingDeps, I'm going to move performance questions like repeated checks for modules to another issue: https://github.com/digital-asset/daml/issues/19859